src/confluent_kafka/src/confluent_kafka.c

```diff
@@ -843,18 +859,31 @@ static int TopicPartition_init (PyObject *self, PyObject *args,
 	const char *topic;
 	int partition = RD_KAFKA_PARTITION_UA;
 	long long offset = RD_KAFKA_OFFSET_INVALID;
+	const char *metadata;
+	size_t metadata_size = 0;
+	static char *kws_optional[] = { "topic",
+					"partition",
+					"offset",
+					"metadata",
+					"metadata_size",
+					NULL };
 	static char *kws[] = { "topic",
 			       "partition",
 			       "offset",
 			       NULL };
-
-	if (!PyArg_ParseTupleAndKeywords(args, kwargs, "s|iL", kws,
-					 &topic, &partition, &offset))
+	if (PyArg_ParseTupleAndKeywords (args, kwargs, "s|iLsL", kws_optional,
```

Comment:
Can't call PyArg_ParseTupleAndKeywords() again on failure since it will have already raised an exception.
All arguments should be parsable with the same call, and since you are just adding metadata that should be fine.

Reply:
I think the metadata type should be `y*` (or `y#`?) for binary, rather than `s` which is unicode or binary.